File: packages/git-core/src/parse/network.ts
/**
 * `git push --porcelain` and `git fetch --porcelain` parsers.
 *
 * These are two different formats that happen to share a name — the design calls
 * this out, and the measured output confirms it:
 *
 * ```
 * push:  To <url>\n*\trefs/heads/main:refs/heads/main\t[new branch]\nDone\n
 *        \trefs/heads/main:refs/heads/main\t49f6bed..5fe8bd1
 *        !\trefs/heads/main:refs/heads/main\t[rejected] (non-fast-forward)
 *        +\trefs/heads/main:refs/heads/main\t5fe8bd1...7c46aef (forced update)
 *        =\trefs/heads/main:refs/heads/main\t[up to date]
 *
 * fetch: <flag> <old-oid> <new-oid> <local-ref>\n    (one line per updated ref)
 *        (empty output when nothing moved)
 * ```
 *
 * Both parsers are tolerant about *extra* lines: a remote may print `remote: …`
 * progress, and a rejected push exits non-zero with the reason on stderr. The
 * per-ref records are what the workflow reports, and unparsed lines are kept as
 * diagnostics so nothing is silently dropped.
 */
import { GitOutputParseError } from "../ports.js";

export type PushRefStatus =
  "upToDate" | "new" | "fastForward" | "forced" | "deleted" | "rejected";

export interface PushRefResult {
  readonly from: string;
  readonly to: string;
  readonly status: PushRefStatus;
  /** Summary as Git printed it: `[new branch]`, `old..new (forced update)`, … */
  readonly summary: string;
}

export interface PushParseResult {
  /** The remote URL/name as Git echoed it, redacted by the host for display. */
  readonly remote: string | null;
  readonly refs: readonly PushRefResult[];
  /** True when Git reported there was nothing to push at all. */
  readonly everythingUpToDate: boolean;
  readonly diagnostics: readonly string[];
}

export interface FetchRefResult {
  readonly flag: string;
  readonly oldOid: string;
  readonly newOid: string;
  /** The local ref that moved, e.g. `refs/remotes/origin/main`. */
  readonly localRef: string;
}

export interface FetchParseResult {
  readonly refs: readonly FetchRefResult[];
  readonly diagnostics: readonly string[];
}

const PUSH_FORMAT = "push --porcelain";

function splitLines(bytes: Uint8Array): Uint8Array[] {
  const lines: Uint8Array[] = [];
  let start = 0;
  for (let index = 0; index < bytes.byteLength; index += 1) {
    if (bytes[index] === 0x0a) {
      lines.push(bytes.subarray(start, index));
      start = index + 1;
    }
  }
  if (start < bytes.byteLength) {
    lines.push(bytes.subarray(start));
  }
  return lines;
}

/**
 * Patch text and porcelain progress are display material, so this decodes as
 * UTF-8 with replacement. It never feeds an execution argument — that is why
 * paths elsewhere stay as bytes.
 */
function decode(bytes: Uint8Array): string {
  let text = "";
  let index = 0;
  while (index < bytes.byteLength) {
    const first = bytes[index] ?? 0;
    if (first < 0x80) {
      text += String.fromCharCode(first);
      index += 1;
      continue;
    }
    let needed = 0;
    let codePoint = 0;
    if ((first & 0xe0) === 0xc0) {
      needed = 1;
      codePoint = first & 0x1f;
    } else if ((first & 0xf0) === 0xe0) {
      needed = 2;
      codePoint = first & 0x0f;
    } else if ((first & 0xf8) === 0xf0) {
      needed = 3;
      codePoint = first & 0x07;
    } else {
      text += "\uFFFD";
      index += 1;
      continue;
    }
    if (index + needed >= bytes.byteLength) {
      text += "\uFFFD";
      break;
    }
    for (let offset = 1; offset <= needed; offset += 1) {
      const continuation = bytes[index + offset] ?? 0;
      codePoint = (codePoint << 6) | (continuation & 0x3f);
    }
    text += String.fromCodePoint(codePoint);
    index += needed + 1;
  }
  return text;
}

export function parsePushPorcelain(bytes: Uint8Array): PushParseResult {
  const refs: PushRefResult[] = [];
  const diagnostics: string[] = [];
  let remote: string | null = null;
  let everythingUpToDate = false;

  for (const lineBytes of splitLines(bytes)) {
    const line = decode(lineBytes);
    if (line.length === 0) {
      continue;
    }
    if (line.startsWith("To ")) {
      remote = line.slice(3);
      continue;
    }
    if (line === "Done") {
      continue;
    }
    if (line.startsWith("Everything up-to-date")) {
      everythingUpToDate = true;
      continue;
    }
    // A ref line is `FLAG TAB from:to TAB summary`, where FLAG is one character
    // and may be a space. Splitting on the first tab in the whole line would take
    // the flag for the refspec, so the position is fixed instead.
    if (line.length > 2 && line[1] === "\t") {
      const flag = line[0] ?? "?";
      const [refspec, summaryPart] = splitOnce(line.slice(2), "\t");
      const [from, to] = splitOnce(refspec, ":");
      if (to.length === 0) {
        throw new GitOutputParseError(
          PUSH_FORMAT,
          `ref line '${line}' has no destination ref`,
        );
      }
      refs.push({
        from,
        to,
        status: pushStatusFromFlag(flag, summaryPart),
        summary: summaryPart,
      });
      continue;
    }
    // Anything else (remote: messages, hints) is diagnostic material.
    diagnostics.push(line);
  }

  return { remote, refs, everythingUpToDate, diagnostics };
}

function pushStatusFromFlag(flag: string, summary: string): PushRefStatus {
  if (summary.startsWith("[rejected]")) {
    return "rejected";
  }
  switch (flag) {
    case " ":
      return "fastForward";
    case "+":
      return "forced";
    case "-":
      return "deleted";
    case "*":
      return "new";
    case "=":
      return "upToDate";
    case "!":
      return "rejected";
    default:
      throw new GitOutputParseError(
        PUSH_FORMAT,
        `unknown push status flag '${flag}'`,
      );
  }
}

function splitOnce(value: string, separator: string): [string, string] {
  const index = value.indexOf(separator);
  if (index === -1) {
    return [value, ""];
  }
  return [value.slice(0, index), value.slice(index + separator.length)];
}

/**
 * Parse `fetch --porcelain`.
 *
 * One line per updated ref: a flag character, then space-separated old object
 * name, new object name and the local ref. Git documents this output as not yet
 * stable, so the parser is strict about the shape it accepts and the test suite
 * regenerates it against the pinned Git version rather than trusting this comment.
 */
export function parseFetchPorcelain(bytes: Uint8Array): FetchParseResult {
  const refs: FetchRefResult[] = [];
  const diagnostics: string[] = [];

  for (const lineBytes of splitLines(bytes)) {
    const line = decode(lineBytes);
    if (line.length === 0) {
      continue;
    }
    // The line is `<flag> <old-oid> <new-oid> <local-ref>` with a single-character
    // flag, which is usually a space — so the line often begins with two spaces and
    // splitting on runs of whitespace would lose the flag.
    if (line.length > 2 && line[1] === " ") {
      const flag = line[0] ?? "?";
      const parts = line.slice(2).split(" ");
      const [oldOid, newOid, localRef] = parts;
      if (
        parts.length === 3 &&
        oldOid !== undefined &&
        newOid !== undefined &&
        localRef !== undefined &&
        isOidOrZero(oldOid) &&
        isOidOrZero(newOid) &&
        localRef.startsWith("refs/")
      ) {
        refs.push({ flag, oldOid, newOid, localRef });
        continue;
      }
    }
    diagnostics.push(line);
  }

  return { refs, diagnostics };
}

function isOidOrZero(value: string): boolean {
  return /^[0-9a-f]{40}$|^[0-9a-f]{64}$/.test(value);
}

File: crates/refyard-core/src/parse/network.rs
//! Parsing `git fetch --porcelain` (and the shared object-name shape) into per-ref
//! outcomes.
//!
//! The porcelain table is one line per ref: `<flag> <old-oid> <new-oid> <local-ref>`.
//! The flag is a *single character* and is usually a space (fast-forward), so the line
//! often begins with two spaces and splitting on runs of whitespace would lose the flag
//! — the parse keys off `line[1] == ' '` instead. A line that does not fit the shape is
//! kept as a diagnostic rather than dropped or guessed at: "what did this fetch touch"
//! is only ever what the bytes show.

/// One ref a fetch named: its flag, the object names before and after, and the local
/// ref it landed in.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FetchRefResult {
    /// The porcelain flag: `=` up to date, ` ` fast-forward, `+` forced, `-` deleted,
    /// `*` created.
    pub flag: u8,
    pub old_oid: String,
    pub new_oid: String,
    pub local_ref: String,
}

impl FetchRefResult {
    /// Did this ref actually move? `=` is "up to date", which changed nothing.
    pub fn moved(&self) -> bool {
        self.flag != b'='
    }
}

/// The whole `--porcelain` table: the refs it named, plus any line that did not fit the
/// format (kept as diagnostics rather than silently discarded).
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct FetchParseResult {
    pub refs: Vec<FetchRefResult>,
    pub diagnostics: Vec<String>,
}

/// A SHA-1 (40 hex) or SHA-256 (64 hex) object name, including the all-zero name Git
/// prints for a created or deleted ref's non-existent end.
fn is_oid_or_zero(value: &str) -> bool {
    (value.len() == 40 || value.len() == 64) && value.bytes().all(|b| b.is_ascii_hexdigit())
}

/// Parse `git fetch --porcelain` stdout. Byte-level: the delimiters are ASCII, and a
/// field is only a field if its bytes are exactly an object name or a `refs/` name.
pub fn parse_fetch_porcelain(bytes: &[u8]) -> FetchParseResult {
    let mut result = FetchParseResult::default();
    for mut line in bytes.split(|b| *b == b'\n') {
        if line.last() == Some(&b'\r') {
            line = &line[..line.len() - 1];
        }
        if line.is_empty() {
            continue;
        }
        // `<flag> <space> …` — the flag is one byte and the separator one space.
        if line.len() > 2 && line[1] == b' ' {
            let flag = line[0];
            let parts: Vec<&[u8]> = line[2..].split(|b| *b == b' ').collect();
            if parts.len() == 3 {
                let old_oid = std::str::from_utf8(parts[0]).unwrap_or("");
                let new_oid = std::str::from_utf8(parts[1]).unwrap_or("");
                let local_ref = std::str::from_utf8(parts[2]).unwrap_or("");
                if is_oid_or_zero(old_oid)
                    && is_oid_or_zero(new_oid)
                    && local_ref.starts_with("refs/")
                {
                    result.refs.push(FetchRefResult {
                        flag,
                        old_oid: old_oid.to_string(),
                        new_oid: new_oid.to_string(),
                        local_ref: local_ref.to_string(),
                    });
                    continue;
                }
            }
        }
        result
            .diagnostics
            .push(String::from_utf8_lossy(line).to_string());
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    const ZERO: &str = "0000000000000000000000000000000000000000";
    const OLD: &str = "1111111111111111111111111111111111111111";
    const NEW: &str = "2222222222222222222222222222222222222222";

    #[test]
    fn the_flag_is_one_byte_even_when_it_is_a_space() {
        // A fast-forward line starts with *two* spaces; splitting on whitespace runs
        // would eat the flag and shift every field. This is the bug the shape guards.
        let table = format!("  {OLD} {NEW} refs/remotes/origin/main\n");
        let parsed = parse_fetch_porcelain(table.as_bytes());
        assert_eq!(parsed.refs.len(), 1, "{:?}", parsed.diagnostics);
        let r = &parsed.refs[0];
        assert_eq!(r.flag, b' ', "the flag is the leading space");
        assert_eq!(r.old_oid, OLD);
        assert_eq!(r.new_oid, NEW);
        assert_eq!(r.local_ref, "refs/remotes/origin/main");
        assert!(r.moved());
    }

    #[test]
    fn each_flag_shape_parses_and_up_to_date_counts_as_unchanged() {
        let table = format!(
            "* {ZERO} {NEW} refs/tags/v1\n+ {OLD} {NEW} refs/heads/main\n- {OLD} {ZERO} refs/heads/gone\n= {OLD} {OLD} refs/heads/same\n"
        );
        let parsed = parse_fetch_porcelain(table.as_bytes());
        assert_eq!(parsed.refs.len(), 4, "{:?}", parsed.diagnostics);
        assert!(parsed.refs[0].moved(), "a created ref moved");
        assert!(parsed.refs[1].moved(), "a forced update moved");
        assert!(parsed.refs[2].moved(), "a deletion moved");
        assert!(!parsed.refs[3].moved(), "`=` is up to date: nothing moved");
    }

    #[test]
    fn a_line_that_is_not_the_shape_is_a_diagnostic_never_a_ref() {
        let table = format!(
            "From https://example.com/r\n * [new branch]      main -> origin/main\n  {OLD} {NEW} refs/remotes/origin/main\n"
        );
        let parsed = parse_fetch_porcelain(table.as_bytes());
        assert_eq!(parsed.refs.len(), 1, "only the well-formed ref line parsed");
        assert_eq!(parsed.diagnostics.len(), 2, "the other two lines are kept");
    }

    #[test]
    fn an_object_name_is_40_or_64_hex_and_nothing_else() {
        assert!(is_oid_or_zero(OLD));
        assert!(is_oid_or_zero(ZERO));
        assert!(is_oid_or_zero(&"a".repeat(64)));
        assert!(!is_oid_or_zero("abc"));
        assert!(!is_oid_or_zero(&"a".repeat(39)));
        assert!(!is_oid_or_zero(&"z".repeat(40)));
    }
}
